### PR TITLE
Video Optimization: Add missing `crossorigin` attributes

### LINKED ISSUE #7016: Video Optimization: 3P Media not loading with cross-origin isolation
## Bug Description

This is a follow-up to #6875 .

With these new headers, 3P media are no longer loaded due to missing `crossorigin` attributes.

## Expected Behaviour

All 3P media (images, videos, GIFs) should load just fine

## Steps to Reproduce

1. Enable video optimization
2. Go to 3P media tab
3. See console errors

## Screenshots

Screenshot 2021-04-04 at 16 36 00

## Additional Context

- Plugin Version: 1.6.0-alpha
- WordPress Version: 5.7
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

All 3P media (images, videos, GIFs) should load just fine when cross-origin isolation is enabled.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

This is a follow-up to #6875 .

With these new headers, 3P media are no longer loaded due to missing crossorigin attributes.

## Relevant Technical Choices

I have also improved poster images, by making sure, loading=lazy, alt and onload are loaded on poster images as well. 

This PR will likely fix another issue, where images from media library are served from another url, like if assets are loaded from s3.

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Enable video optimization experiment
2. Browsing and using 3P Media (videos, images, GIFs) should work without issues or console errors  
    **Note:** there can be some console errors for GIFs due to caching, but most of them should load fine already.
3. Toggling "show images" in Page templates tab should not cause console errors

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7016

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js [hunk 1]
```diff
@@ -134,20 +137,31 @@ function InnerElement({
...
   };
-  const cloneProps = {
+
+  if (src && !src.startsWith(siteUrl)) {
+    commonProps.crossOrigin = 'anonymous';
+  }
+
+  const commonImageProps = {
     ...commonProps,
+    onLoad: makeMediaVisible,
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: Why was `onLoad` moved from `imageProps` to the common props here? 🤔

**Comment by swissspidy** [comment]
Didn't realize you picked this up already! Would have been good to discuss possible implementations beforehand. But kudos for finding possible points of conflict.

I'm not really fond of sprinking `useConfig()` calls  and site URL comparisons all over the place.
If anything, couldn't we just compare with `window.location.host`?

Also, could the attribute perhaps simply be added by default?

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: > I have also improved poster images, by making sure, loading=lazy, alt and onload are loaded on poster images as well.

**Comment by spacedmonkey** [comment]
> Didn't realize you picked this up already! Would have been good to discuss possible implementations beforehand. But kudos for finding possible points of conflict.

I picked it up as it is a P0, which is drop everything.

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: This just makes the image props between poster image and image the same.

**Comment by spacedmonkey** [comment]
> I'm not really fond of sprinking useConfig() calls and site URL comparisons all over the place.
If anything, couldn't we just compare with window.location.host?

I think that window.location.origin might be better, but I could change that. Would this have any issues related to tests and using window?

> Also, could the attribute perhaps simply be added by default?

I could do this, but image tags are used in LOTs of places and I didn't want to add it unless it is needed.

**Comment by swissspidy** [comment]
> > I'm not really fond of sprinking useConfig() calls and site URL comparisons all over the place.
> > If anything, couldn't we just compare with window.location.host?
> 
> I think that window.location.origin might be better, but I could change that. Would this have any issues related to tests and using window?

Ah, yes, even better. Let's change it to that then.

> 
> > Also, could the attribute perhaps simply be added by default?
> 
> I could do this, but image tags are used in LOTs of places and I didn't want to add it unless it is needed.

I didn't mean changing _all_ image tags. Just in the 8 files your PR currently touches. So just removing the comparison.

File: assets/src/web-stories-block/block/block-types/story-embed-block/embedControls.js [hunk 2]
```diff
@@ -128,7 +128,7 @@ const EmbedControls = (props) => {
                 <BaseControl.VisualLabel className="web-stories-embed-poster-label">
                   {__('Poster image', 'web-stories')}
                 </BaseControl.VisualLabel>
-                {hasPoster && <img src={poster} alt="" />}
+                {hasPoster && <img src={poster} alt="" crossOrigin="anonymous" />}
```
Comment (spacedmonkey) [inline assets/src/web-stories-block/block/block-types/story-embed-block/embedControls.js]:
```suggestion
                {hasPoster && <img src={poster} alt="" />}
```

**Comment by spacedmonkey** [comment]
Changed the PR, so that is just load cross origin on video and image tags everywhere

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js (hunk 1 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: But I don't think the poster needs an `onLoad` callback?

Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: The onload just fades in the image. Poster image are image tags like an image, they should fail in the same way a image does.

Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:144]: Also It already has a onload- https://github.com/google/web-stories-wp/pull/7086/files#diff-156fd9c14d28faad8ca9b94fd115f288b4a1511b74ac5b3c2bdd5db277f0ab91L193 This isnt a change.

**Comment by swissspidy** [comment]
Pushed 90ecd994623a6390d524fa0a5f6802fafd2adbac to fix some [issues with the E2E tests](https://github.com/google/web-stories-wp/runs/2280259101#step:15:9) I was seeing

**Comment by spacedmonkey** [comment]
@swissspidy I think that issues are valid, the assets are getting a cors error. What are the cor policy on wp.stories.google? 

> 'https://wp.stories.google/static/main//images/templates/beauty/beauty_page9_story3.jpg' from origin 'http://localhost:8899' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource."],["Failed to load resource: net::ERR_FAILED"]

**Comment by swissspidy** [comment]
> @swissspidy I think that issues are valid, the assets are getting a cors error. What are the cor policy on wp.stories.google?
> 
> > '[wp.stories.google/static/main//images/templates/beauty/beauty_page9_story3.jpg](https://wp.stories.google/static/main//images/templates/beauty/beauty_page9_story3.jpg)' from origin 'http://localhost:8899' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource."],["Failed to load resource: net::ERR_FAILED"]

Well yes it was valid, but that change fixes it.

The issue is the double slashes in `wp.stories.google/static/main//images/templates/beauty/beauty_page9_story3.jpg`, which causes a 301 to `wp.stories.google/static/main/images/templates/beauty/beauty_page9_story3.jpg` (one slash).

The 2-slash URL doesn't have the CORS header, the 1-slash URL does.

**Comment by spacedmonkey** [comment]
Screenshot 2021-04-07 at 10 51 39
Removing the cross origin attribute on quick tips. It seems to work.

**Comment by spacedmonkey** [comment]
@choumx @swissspidy Any chance of re-review?

**Comment by spacedmonkey** [comment]
Screenshot 2021-04-09 at 11 20 40
Confirmed templates images are loading.

**Comment by swissspidy** [comment]
> Screenshot 2021-04-07 at 10 51 39
> 
> Removing the cross origin attribute on quick tips. It seems to work.

It doesn't in my testing. Can you revert this change again?

**Comment by swissspidy** [comment]
Two major issues I see:

1. Loading GIFs from Tenor in the 3P Media tab does not work.  
  Looks like Tenor doesn't add an `Access-Control-Allow-Origin` header for requests to `c.tenor.com`, even if an `Origin` header is present.  
  To work around this, it seems like we **must not** add the `crossorigin` attribute to these images.  
  In the meantime we could perhaps look into whether this is intentional on the Tenor side or if the header could be added. (cc @choumx)
2. ~There are console errors when going to the Page Templates tab and clicking on "Show images"  
  Looks like it's because it triggers the `preloadImage` function, which does not set the `crossorigin` attribute. So we need to set it there.~  
   Fixed by 359ba0a7e872f3a48ceb88fab9045eed335bee9f

Related: we should really improve error handling around `preloadImage`. It can return a rejected Promise, but it's never caught. We might want to update `preloadImage` to just never `reject`, but instead `resolve(null)` & log errors or something. Or fix all usage to wrap calls in `try {} catch {}`

**Comment by dreamofabear** [comment]
> In the meantime we could perhaps look into whether this is intentional on the Tenor side or if the header could be added.

FYI following up with the Tenor team on this. Looks like it should be fixable on their end, perhaps as early as next week.

**Comment by dreamofabear** [comment]
@barklund I think all these `crossorigin` media changes should make our render-to-canvas features (e.g. eyedropper, text contrast checks, text magic) work better by avoiding canvas tainting.

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js [hunk 3]
```diff
@@ -136,20 +136,28 @@ function InnerElement({
...
-  const cloneProps = {
+
+  const commonImageProps = {
     ...commonProps,
+    onLoad: makeMediaVisible,
     loading: 'lazy',
+  };
+
+  const cloneProps = {
+    ...commonImageProps,
```
Comment (miina) [inline assets/src/edit-story/components/library/panes/media/common/innerElement.js:150]: > This just makes the image props between poster image and image the same.

@spacedmonkey `cloneProps` shouldn't really need `onLoad`, what was the reason for adding it there as well? Clone is the visible clone image while dragging media from the library onto a page, it only becomes visible once starting dragging an image/video. Thoughts?

**Comment by swissspidy** [comment]
> FYI following up with the Tenor team on this. Looks like it should be fixable on their end, perhaps as early as next week.

Would be good to have an ETA. If we omit the `crossorigin` attribute  for Tenor images for now, and they add the header later on, we suddenly need to push a release to add back `crossorigin`

File: assets/src/edit-story/components/library/panes/media/common/innerElement.js (hunk 3 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/common/innerElement.js:150]: Definitely looks like a bug. Reverted in d7b236efb7ba2f7515509d9e2afe11ad791b6882

**Comment by swissspidy** [comment]
> > FYI following up with the Tenor team on this. Looks like it should be fixable on their end, perhaps as early as next week.
> 
> Would be good to have an ETA. If we omit the `crossorigin` attribute for Tenor images for now, and they add the header later on, we suddenly need to push a release to add back `crossorigin`

I quickly tried omitting this, but it meaans that **everywhere** we add `crossorigin=anonymous` (library, canvas, etc) we would have to do a check whether the URL is from Tenor. Not really feasible.

**Comment by swissspidy** [comment]
@choumx IMO the Tenor change is quite the blocker here and we should wait until that's addressed. WDYT?

**Comment by dreamofabear** [comment]
Yea let's wait for the Tenor fix for this PR. I'll follow up on a ETA today.

**Comment by dreamofabear** [comment]
The Tenor issue is now fixed and rolled out. I still see some requests to `c.tenor.com` throwing this error due to proxy caching, but should be resolved by tomorrow.

**Review by swissspidy — APPROVED** [review]

**Pull request merged**